Frictional contact damper

bwspenc opened the pull request:
ref #6524

This PR has two commits: the commit that is in currently active PR #6528, and a second commit that adds a damper to limit slip reversals for frictional contact.

I'm submitting this now because I'd like to start a discussion on how to do this.  We shouldn't actually merge this until the other PR gets merged and I remove that first commit.

This damping capability is currently implemented in the `updateSolution()` method in `FrictionalContactDamperProblem` which derives from `Problem`.  It works, but I think this is ugly.  I'd much rather do this in a `Damper`.  The problem is that the current Damper objects are only called within an element loop.  I think that the right way to do this is to expand the current Damper system to include GeneralDampers and ElementDampers, just like UserObjects. My code here would go in a General Damper.  Thoughts? How big of a development would that be?

permcody commented:
I guess a bigger question would be: Do `Dampers` in general need to be as flexible as `UserObjects`? Are there cases where we need per element values, per node values, or general cases?

That'll help us decide which design to work towards. If it turns out we want that flexibility perhaps the Dampers should inherit from the UserObject system instead of being stand alone. I think @aeslaughter's `Warehouse` redesign will make this fairly easy to accomplish. 

@idaholab/moose-developers please comment

friedmud commented:
I vote for Dampers being UserObjects.

Derek

On Tue, Mar 8, 2016 at 6:22 PM Cody Permann <notifications@github.com>
wrote:

> I guess a bigger question would be: Do Dampers in general need to be as
> flexible as UserObjects? Are there cases where we need per element
> values, per node values, or general cases?
>
> That'll help us decide which design to work towards. If it turns out we
> want that flexibility perhaps the Dampers should inherit from the
> UserObject system instead of being stand alone. I think @aeslaughter
> <https://github.com/aeslaughter>'s Warehouse redesign will make this
> fairly easy to accomplish.
>
> @idaholab/moose-developers
> <https://github.com/orgs/idaholab/teams/moose-developers> please comment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6529#issuecomment-194015049>.
>

friedmud commented:
To be clear: Dampers should be their own System still... but should inherit
from UserObjects and have the same traits.

On Tue, Mar 8, 2016 at 6:31 PM Derek Gaston <friedmud@gmail.com> wrote:

> I vote for Dampers being UserObjects.
>
> Derek
>
> On Tue, Mar 8, 2016 at 6:22 PM Cody Permann <notifications@github.com>
> wrote:
>
>> I guess a bigger question would be: Do Dampers in general need to be as
>> flexible as UserObjects? Are there cases where we need per element
>> values, per node values, or general cases?
>>
>> That'll help us decide which design to work towards. If it turns out we
>> want that flexibility perhaps the Dampers should inherit from the
>> UserObject system instead of being stand alone. I think @aeslaughter
>> <https://github.com/aeslaughter>'s Warehouse redesign will make this
>> fairly easy to accomplish.
>>
>> @idaholab/moose-developers
>> <https://github.com/orgs/idaholab/teams/moose-developers> please comment
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/idaholab/moose/pull/6529#issuecomment-194015049>.
>>
>

aeslaughter commented:
If we are talking about making Dampers UserObjects I think we should be talking about starting a more generic object structure so we can move towards complete dependency resolution.

Why not re-factor dampers, arguably are simplest system, to new generic set of objects.
* NodalMooseObject -> MooseObject
* ElementMooseObject -> MooseObject

We can then create the Dampers as:
NodalDamper->NodalMooseObject
                      ->Damper

ElementDamper->ElementMooseObject
                          ->Damper

Then FEProblem::addDamper would:
  * Add the "damper" to the correct Nodal/ElementMooseObject warehouse and
  * Add the "damper" to the damper warehouse for calls to getDamperWhateverTheNameIsValueFunction stuff that we do now.

Then create Nodal/ElementMooseObjectComputeThread objects that simply take a std::vector<MooseSharedPointer<Nodal/ElementMooseObject> >. Thus, the dependency resolution can create a vector of these thread objects based on the get method overloading. Finally, you iterate over this vector of thread objects and do the thread reduce stuff.

Then as we go we can migrate other stuff into these more generic containers. This really would be all that difficult to do in Dampers and I prefer this to bolting these into UserObjects which basically follow this idea already, we just need to make it more general.

permcody commented:
We need to gather a few more requirements before we go this direction. One of the key issues why we don't do dependency resolution among some of these mesh entity based objects is because we don't know whether or not they need to be fully reduced before being used or can be used per entity. The postprocessor types often need to fully execute before they can be used in a meaningful way. We should be able to deal with this if we have the right design. There maybe a certain flag that gets set or perhaps an object type that allows "interleaved" execution versus "multiple loop" execution.

Before anyone goes off and starts implementing something, let's continue this discussion and make sure we get the design correct. @YaqWang has a few requirements with resolution among Postprocessors and AuxKernels that he'd like to see handled in the future with additional levels of dependency resolution among Postprocessors and AuxKernels (#3206, #3088).

Regarding design we also need to look at what we've discussed on #1050. 

This is all a little off from the main task at hand though. We have other bigger priorities right now that need to get resolved before we start on this project. We should be able to help @bwspenc in the meantime without too much fuss.

friedmud commented:
Agree with @permcody... and I actually want to retract my statement about
Dampers inheriting from UserObjects.  Dampers are _not_ UserObjects.  We
simply need to break out Dampers into Nodal/Elemental/Side just like we do
for ErrorEstimators, etc.  We don't have to do all of them right now.

@bwspenc : if you just want to implement the one you need right now that
would be fine.

On Wed, Mar 9, 2016 at 10:01 AM Cody Permann <notifications@github.com>
wrote:

> We need to gather a few more requirements before we go this direction. One
> of the key issues why we don't do dependency resolution among some of these
> mesh entity based objects is because we don't know whether or not they need
> to be fully reduced before being used or can be used per entity. The
> postprocessor types often need to fully execute before they can be used in
> a meaningful way. We should be able to deal with this if we have the right
> design. There maybe a certain flag that gets set or perhaps an object type
> that allows "interleaved" execution versus "multiple loop" execution.
>
> Before anyone goes off and starts implementing something, let's continue
> this discussion and make sure we get the design correct. @YaqWang
> <https://github.com/YaqWang> has a few requirements with resolution among
> Postprocessors and AuxKernels that he'd like to see handled in the future
> with additional levels of dependency resolution among Postprocessors and
> AuxKernels (#3206 <https://github.com/idaholab/moose/issues/3206>, #3088
> <https://github.com/idaholab/moose/issues/3088>).
>
> Regarding design we also need to look at what we've discussed on #1050
> <https://github.com/idaholab/moose/issues/1050>.
>
> This is all a little off from the main task at hand though. We have other
> bigger priorities right now that need to get resolved before we start on
> this project. We should be able to help @bwspenc
> <https://github.com/bwspenc> in the meantime without too much fuss.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6529#issuecomment-194334524>.
>

bwspenc commented:
I think the near-term path forward is to create a GeneralDamper (which is what I need) and an ElementDamper (which is what the current Damper is).  I don't see an immediate need for a NodalDamper, but we could certainly add that down the road. I'd be happy to take a stab at implementation of that. When I'm ready to do that, I'll come by and talk about the architecture.

bwspenc committed:
Pass solution and update into damper ref #6524

This is needed for the ContactSlipDamper, which I will add in a
subsequent commit.

bwspenc committed:
Add ContactSlipDamper ref #6524

bwspenc commented:
@permcody I just pushed new commits into this branch that implement the Damper instead of the original version. This new design is the way I believe it should be done. If it's better to close this PR and create a new clean one, I can do that.

permcody commented:
No need to open another PR. I'll take a look.

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+  params.addRequiredParam<std::vector<int> >("slave", "IDs of the slave surfaces for which slip reversals should be damped");
+  params.addParam<Real>("max_iterative_slip", std::numeric_limits<Real>::max(), "Maximum iterative slip");
+  params.addParam<Real>("min_damping_factor", 0.0, "Minimum permissible value for damping factor");
+  params.addParam<Real>("damping_threshold_factor", 1.0e3, "If previous iterations's slip is below the slip tolerance, only damp a slip reversal if the slip magnitude is greater than than this factor times the old slip.");
```
Should this be 1.0e-3?

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+  params.addRequiredParam<std::vector<int> >("master", "IDs of the master surfaces for which slip reversals should be damped");
+  params.addRequiredParam<std::vector<int> >("slave", "IDs of the slave surfaces for which slip reversals should be damped");
+  params.addParam<Real>("max_iterative_slip", std::numeric_limits<Real>::max(), "Maximum iterative slip");
+  params.addParam<Real>("min_damping_factor", 0.0, "Minimum permissible value for damping factor");
```
Might use `checkedRangeParam` just to ward off another potential user bug. There may be a few easy opportunities here for checkedRanges.

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+  if (num_interactions == 0)
+    mooseError("Must define at least one master/slave pair in ContactSlipDamper");
+
+  for (unsigned int i=0; i<master.size(); ++i)
```
spacing around `=`

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+  if (num_interactions != slave.size())
+    mooseError("Sizes of master surface and slave surface lists must match in ContactSlipDamper");
+  if (num_interactions == 0)
+    mooseError("Must define at least one master/slave pair in ContactSlipDamper");
```
checkedRange

bwspenc replied:
I'm not quite sure what you mean here.

permcody replied:
Disregard

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+    {
+      std::vector<dof_id_type> & slave_nodes = pen_loc._nearest_node._slave_nodes;
+
+      for (unsigned int i=0; i<slave_nodes.size(); i++)
```
spacing

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+    {
+      std::vector<dof_id_type> & slave_nodes = pen_loc._nearest_node._slave_nodes;
+
+      for (unsigned int i=0; i<slave_nodes.size(); i++)
```
spacing

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+
+  for (unsigned int i=0; i<master.size(); ++i)
+  {
+    std::pair<int,int> ms_pair(master[i],slave[i]);
```
spacing around both commas

permcody commented on framework/include/base/NonlinearSystem.h:
```diff
@@ -262,10 +262,12 @@ class NonlinearSystem : public SystemTempl<TransientNonlinearImplicitSystem>,
    * @return returns The damping factor
    */
-  Real computeDamping(const NumericVector<Number>& update);
+  Real computeDamping(const NumericVector<Number>& solution,
```
spacing around `&`

permcody commented on modules/contact/src/MechanicalContactConstraint.C:
```diff
@@ -277,7 +277,7 @@ MechanicalContactConstraint::computeContactForce(PenetrationInfo * pinfo)
           RealVectorValue contact_force_tangential( pinfo->_contact_force - contact_force_normal );
 
-          RealVectorValue tangential_inc_slip_prev_iter = pinfo->_incremental_slip_prev_iter - (pinfo->_incremental_slip_prev_iter * pinfo->_normal) * pinfo->_normal;
+//          RealVectorValue tangential_inc_slip_prev_iter = pinfo->_incremental_slip_prev_iter - (pinfo->_incremental_slip_prev_iter * pinfo->_normal) * pinfo->_normal;
```
Maybe just remove this?

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+{
+  bool operate_on_this_interaction = false;
+  std::set<std::pair<int,int> >::iterator ipit;
+  std::pair<int,int> ms_pair(pen_loc._master_boundary,pen_loc._slave_boundary);
```
spacing around commas

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+           << std::setw(11) << _num_stick_locked
+           << std::setw(10) << _num_slip_reversed
+           << "\n\n";
+    _console<<std::flush;
```
spacing

permcody commented on modules/contact/include/ContactSlipDamper.h:
```diff
@@ -0,0 +1,70 @@
+  int _num_stick_locked;
+  int _num_slip_reversed;
+  Real _inc_slip_norm;
+  Real _it_slip_norm;
```
I believe both of these Reals are unused - please remove.

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+    _communicator.sum(_num_slipping_friction);
+    _communicator.sum(_num_stick_locked);
+    _communicator.sum(_num_slip_reversed);
+    _communicator.min(damping);
```
This is fine for now but we can consolidate all these into a single communication later if necessary.

bwspenc replied:
I agree.  I'll leave as-is now if that's OK.

permcody commented on modules/contact/src/ContactSlipDamper.C:
```diff
@@ -0,0 +1,226 @@
+        }
+      }
+    }
+    _console<<std::flush;
```
spacing

permcody commented on framework/include/dampers/ConstantDamper.h:
```diff
@@ -39,7 +39,8 @@ class ConstantDamper : public GeneralDamper
    */
-  virtual Real computeDamping(const NumericVector<Number> & update);
+  virtual Real computeDamping(const NumericVector<Number> & solution,
+                              const NumericVector<Number> & update);
```
Please comment out the parameters to avoid debug compile warnings

bwspenc replied:
I think I've always just commented them out in the .C files, which is what I've done here.  I'm happy to comment them out in the .h files, but I don't think it's necessary to avoid compile warnings because the parameter names in the .h files aren't actually used for anything.

permcody replied:
Sorry, was getting carried away. Leave the header alone.

permcody commented on modules/combined/tests/frictional_contact/single_point_2d/tests:
```diff
@@ -4,13 +4,28 @@
+    exodiff = 'single_point_2d_fcp_predictor_out.e'
+    custom_cmp = 'single_point_2d.cmp'
+    superlu = true
     max_parallel = 1
```
Do we need this parallel restriction now that you are using superlu?

bwspenc replied:
We actually do need that because there's a parallel bug in that code (which is the old method I'm replacing with this new work). I'll make an issue for that bug.

permcody commented on modules/contact/include/ContactSlipDamper.h:
```diff
@@ -0,0 +1,70 @@
+   */
+  bool operateOnThisInteraction(const PenetrationLocator & pen_loc);
+
+  std::set<std::pair<int,int> > _interactions;
```
spacing

permcody commented:
Looks good - mostly cosmetic issues.

aeslaughter commented:
The `NonlinearSystem::reinitIncrementForDampers(THREAD_ID tid)` is not being tested, does this PR add a test for this? Also, in this function you need to use the `tid` as `getActiveObjects(tid)` in this function.

bwspenc committed:
Test ContactSlipDamper ref #6524

bwspenc commented:
@aeslaughter That method is probably not being tested because the only ElementDamper is MaxIncrement, which is probably a little hard to test reliably. Thanks for pointing out that issue with tid as well.  Those issues are unrelated to this, so I'll open up a new issue for them.

aeslaughter commented:
@bwspenc Are you ready for this to merge?

bwspenc commented:
@aeslaughter Yes, as far as I'm concerned, it's ready to go.
